libtransmission/global-ip-cache.h

```diff
@@ -0,0 +1,123 @@
+// This file Copyright © 2023-2023 Mnemosyne LLC.
+// It may be used under GPLv2 (SPDX: GPL-2.0-only), GPLv3 (SPDX: GPL-3.0-only),
+// or any future license endorsed by Mnemosyne LLC.
+// License text can be found in the licenses/ folder.
+
+#include <array>
+#include <atomic>
+#include <condition_variable>
+#include <chrono> // operator ""min, operator ""s, std::chrono::milliseconds
+#include <memory> // std::unique_ptr
+#include <mutex>
+#include <optional>
+#include <shared_mutex>
+#include <string_view> // operator ""sv
+
+#include "net.h"
+#include "timer.h"
+#include "tr-assert.h"
+#include "web.h"
+
+#pragma once
+
+#ifndef __TRANSMISSION__
+#error only libtransmission should #include this header.
+#endif
+
+using namespace std::literals;
+
+struct tr_session;
+
+/**
+ * Cache global IP addresses.
+ *
+ * This class caches 3 useful info:
+ * 1. Whether your machine supports the IP protocol
+ * 2. Source address used for global connections
+ * 3. Global address
+ *
+ * The idea is, if this class successfully cached a source address, that means
+ * you have connectivity to the public internet. And if the global address is
+ * the same with the source address, then you are not behind an NAT.
+ *
+ * Note: This class isn't meant to be accessed by anyone other than tr_session,
+ * so it has no public methods.
+ */
+class tr_global_ip_cache
+{
+public:
+    template<typename T>
+    using array_ip_t = std::array<T, NUM_TR_AF_INET_TYPES>;
+
+    explicit tr_global_ip_cache(tr_session& session_in);
+    ~tr_global_ip_cache();
+    tr_global_ip_cache(tr_global_ip_cache const&) = delete;
+    tr_global_ip_cache(tr_global_ip_cache&&) = delete;
+    tr_global_ip_cache& operator=(tr_global_ip_cache const&) = delete;
+    tr_global_ip_cache& operator=(tr_global_ip_cache&&) = delete;
+
+    bool try_shutdown() noexcept;
+
+    [[nodiscard]] std::optional<tr_address> global_addr(tr_address_type type) const noexcept;
+    [[nodiscard]] std::optional<tr_address> global_source_addr(tr_address_type type) const noexcept;
+
+    [[nodiscard]] tr_address bind_addr(tr_address_type type) const noexcept;
```

Comment:
This function makes sense here in the cache but it's slightly awkward that both this function and  tr_session::publicAddress() have to use tr_address::any_ipv4() and ::any_ipv6().

What if this function had the same return type as `tr_session::publicAddress()` and that function just acted as a pass-through to this one?

I haven't looked to see how many callers `bind_addr()` has, so weigh this suggestion against whether or not it makes other code more complicated by making use of this function a little more complex.

Reply:
I think it is fine the way it is, these 2 functions use `any_ipv*()` in different ways, and their logic is also not exactly the same.

Actually, IMO the better way to fix this "awkwardness" is to add member functions to `tr_address` that tests if it is `INADDR_ANY`. Then we can also do away with `tr_session::PublicAddressResult`.

Reply:
I like that idea a lot. :+1:

Reply:
Let's add the `tr_address` methods in another PR. I suppose we don't want too many items in a single PR as it will make changes harder to trace.

Reply:
https://github.com/transmission/transmission/pull/5398